client/extensions/woocommerce/app/settings/email/mailchimp/setup-mailchimp.js

```diff
@@ -0,0 +1,329 @@
+	submitMailchimpNewsletterSettings
+} from 'woocommerce/state/sites/settings/email/actions.js';
+
+const LOG_INTO_MAILCHIMP_STEP = 'log_into';
+const KEY_INPUT_STEP = 'key_input';
+const STORE_INFO_STEP = 'store_info';
+const CAMPAIGN_DEFAULTS_STEP = 'campaign_defaults';
+const NEWSLETTER_SETTINGS_STEP = 'newsletter_settings';
+const STORE_SYNC = 'sync';
+
+const steps = {
+	[ LOG_INTO_MAILCHIMP_STEP ]: { number: 0, nextStep: KEY_INPUT_STEP },
+	[ KEY_INPUT_STEP ]: { number: 1, nextStep: STORE_INFO_STEP },
+	[ STORE_INFO_STEP ]: { number: 2, nextStep: CAMPAIGN_DEFAULTS_STEP },
+	[ CAMPAIGN_DEFAULTS_STEP ]: { number: 3, nextStep: NEWSLETTER_SETTINGS_STEP },
+	[ NEWSLETTER_SETTINGS_STEP ]: { number: 4, nextStep: STORE_SYNC },
+	[ STORE_SYNC ]: { number: 5, nextStep: null },
+};
+
+const storeSettingsRequiredFields = [ 'store_name', 'store_street', 'store_city', 'store_state',
+	'store_postal_code', 'store_country', 'store_phone', 'store_locale', 'store_timezone',
+	'store_currency_code', 'admin_email' ];
+
+const campaignDefaultsRequiredFields = [ 'campaign_from_name', 'campaign_from_email', 'campaign_subject',
+		'campaign_language', 'campaign_permission_reminder' ];
+
+class MailChimpSetup extends React.Component {
+
+	constructor( props ) {
+		super( props );
+		// make this react to the real phase the execution is.
+		this.state = {
+			step: LOG_INTO_MAILCHIMP_STEP,
+			settings: this.prepareDefaultValues( this.props ),
+			settings_values_missing: false,
+			api_key_input: this.props.settings.mailchimp_api_key || '',
+		};
+	}
+
+	componentWillReceiveProps( nextProps ) {
+		// No state changes while doing request.
+		if ( nextProps.isBusy ) {
+			return;
+		}
+
+		const { active_tab } = nextProps.settings;
+		if ( steps[ this.state.step ].nextStep === active_tab ) {
+			const settings = this.prepareDefaultValues( nextProps );
+			this.setState( { step: active_tab, settings } );
+			if ( active_tab === STORE_SYNC ) {
+				nextProps.onClose();
+			}
+		}
+
+		// Update settings if we have received lists
+		if ( nextProps.settings.mailchimp_lists && ! this.props.settings.mailchimp_lists ) {
+			const settings = this.prepareDefaultValues( nextProps );
+			this.setState( { step: active_tab, settings } );
+		}
+	}
+
+	prepareDefaultValues( nextProps ) {
+		const { settings } = nextProps;
+		const newSettings = Object.assign( {}, settings );
+		newSettings.campaign_from_name = settings.campaign_from_name || settings.store_name || '';
+		newSettings.campaign_from_email = settings.campaign_from_email || settings.admin_email || '';
+		newSettings.campaign_subject = settings.campaign_subject || settings.store_name || '';
+		newSettings.store_locale = settings.store_locale || 'en';
+		newSettings.campaign_language = settings.campaign_language || settings.store_locale;
+		newSettings.campaign_permission_reminder = settings.campaign_permission_reminder ||
+			this.props.translate( 'You were subscribed to the newsletter from ' ) + settings.store_name;
```

Comment:
We'll need to use the arg approach instead, especially because of eventual RTL support needs